akka-typed/src/main/scala/akka/typed/ScalaDSL.scala

```diff
@@ -0,0 +1,445 @@
+
+  /**
+   * This type of behavior allows to handle all incoming messages within
+   * the same user-provided partial function, be that a user message or a system
+   * signal. For messages that do not match the partial function the same
+   * behavior is emitted without change. This does entail that unhandled
+   * failures of child actors will lead to a failure in this actor.
+   *
+   * For the lifecycle notifications pertaining to the actor itself this
+   * behavior includes a fallback mechanism: an unhandled [[PreRestart]] signal
+   * will terminate all child actors (transitively) and then emit a [[PostStop]]
+   * signal in addition, whereas an unhandled [[PostRestart]] signal will emit
+   * an additional [[PreStart]] signal.
+   */
+  final case class Full[T](behavior: PartialFunction[MessageOrSignal[T], Behavior[T]]) extends Behavior[T] {
+    override def management(ctx: ActorContext[T], msg: Signal): Behavior[T] = {
+      lazy val fallback: (MessageOrSignal[T]) ⇒ Behavior[T] = {
+        case Sig(context, PreRestart(_)) ⇒
+          context.children foreach { child ⇒
+            context.unwatch(child.untypedRef)
+            context.stop(child)
+          }
+          behavior.applyOrElse(Sig(context, PostStop), fallback)
+        case Sig(context, PostRestart(_)) ⇒ behavior.applyOrElse(Sig(context, PreStart), fallback)
+        case _                            ⇒ Unhandled
+      }
+      behavior.applyOrElse(Sig(ctx, msg), fallback)
+    }
+    override def message(ctx: ActorContext[T], msg: T): Behavior[T] = {
+      behavior.applyOrElse(Msg(ctx, msg), unhandledFunction)
+    }
+    override def toString = s"Full(${LineNumbers(behavior)})"
+  }
+
+  /**
+   * This type of behavior expects a total function that describes the actor’s
+   * reaction to all system signals or user messages, without providing a
+   * fallback mechanism for either. If you use partial function literal syntax
+   * to create the supplied function then any message not matching the list of
+   * cases will fail this actor with a [[scala.MatchError]].
+   */
+  final case class FullTotal[T](behavior: MessageOrSignal[T] ⇒ Behavior[T]) extends Behavior[T] {
+    override def management(ctx: ActorContext[T], msg: Signal) = behavior(Sig(ctx, msg))
+    override def message(ctx: ActorContext[T], msg: T) = behavior(Msg(ctx, msg))
+    override def toString = s"FullTotal(${LineNumbers(behavior)})"
+  }
+
+  /**
+   * This type of behavior is created from a total function from the declared
+   * message type to the next behavior, which means that all possible incoming
+   * messages for the given type must be handled. All system signals are
+   * ignored by this behavior, which implies that a failure of a child actor
+   * will be escalated unconditionally.
+   *
+   * This behavior type is most useful for leaf actors that do not create child
+   * actors themselves.
+   */
+  final case class Total[T](behavior: T ⇒ Behavior[T]) extends Behavior[T] {
```

Comment:
The basic rule is: use whatever works. Using the “wrong” one will be very obvious, because it does not allow you to do what you want to do. Using an overly powerful version will just make the code a little more verbose, so no harm done.